impl/src/main/java/org/jboss/weld/manager/BeanManagerImpl.java

```diff
@@ -1559,4 +1569,160 @@ public void validationFailed(Exception failure, Environment environment) {
         }
     }
 
+    private void removeUnusedBeans() {
+        String excludeTypeProperty = getServices().get(WeldConfiguration.class).getStringProperty(ConfigurationKey.UNUSED_BEANS_EXCLUDE_TYPE);
+
+        if (!excludeTypeProperty.equals(ConfigurationKey.UNUSED_BEANS_EXCLUDE_TYPE.getDefaultValue())) {
+            String excludeAnnotationProperty = getServices().get(WeldConfiguration.class).getStringProperty(ConfigurationKey.UNUSED_BEANS_EXCLUDE_ANNOTATION);
+            // Init exclude patterns
+            Pattern excludeAnnotation = excludeAnnotationProperty.isEmpty() ? null : Pattern.compile(excludeAnnotationProperty);
+            Pattern excludeType = excludeTypeProperty.isEmpty() ? null : Pattern.compile(excludeTypeProperty);
+            Validator validator = getServices().get(Validator.class);
+            // Build bean to declared producers and declared observers maps
+            SetMultimap<Bean<?>, AbstractProducerBean<?, ?, ?>> beanToDeclaredProducers = SetMultimap.newSetMultimap();
+            SetMultimap<Bean<?>, ObserverMethod<?>> beanToDeclaredObservers = SetMultimap.newSetMultimap();
+            for (Bean<?> bean : enabledBeans) {
+                if (bean instanceof AbstractProducerBean) {
+                    AbstractProducerBean<?, ?, ?> producer = (AbstractProducerBean<?, ?, ?>) bean;
+                    beanToDeclaredProducers.put(producer.getDeclaringBean(), producer);
+                }
+            }
+            for (ObserverMethod<?> observerMethod : observers) {
+                if (observerMethod instanceof ObserverMethodImpl) {
+                    ObserverMethodImpl<?, ?> observerMethodImpl = (ObserverMethodImpl<?, ?>) observerMethod;
+                    beanToDeclaredObservers.put(observerMethodImpl.getDeclaringBean(), observerMethod);
+                }
+            }
+            Set<Bean<?>> removable = new HashSet<>();
+            Set<Bean<?>> unusedProducers = new HashSet<>();
+            WeldUnusedMetadataExtension metadataExtension = getUnusedMetadataExtension();
+
+            for (Bean<?> bean : enabledBeans) {
```

Comment:
We are going to be looping three times over all enabled beans.
That sounds a but pricy in huge deployments (and I think we have cleanup auto-enabled in WFLY, don't we?).
Not that I would know how to make it better yet, just thinking aloud :)